### PR TITLE
Fix mathtext mismatched braces

### LINKED ISSUE #26510: [Bug]: mathtext silently ignores content after mismatched opening brace
### Bug summary

All's in the title.

### Code for reproduction

```python
from pylab import *; figtext(.5, .5, r"$a={a$")
```

### Actual outcome

"a = " is printed.

### Expected outcome

An exception is thrown.

### Additional information

This appears at least as far back as 3.5, so is unlikely to come from the recent mathtext work.

### Operating system

macos

### Matplotlib Version

3.8.0.dev1801+g5d5700a99e

### Matplotlib Backend

any

### Python version

3.11.3

### Jupyter version

_No response_

### Installation

git checkout

### PR DESCRIPTION
**Pull request by devRD**
## PR summary
Fixes #26510

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by ksunden** [comment]
We discussed this a bit on GSOC call, the error message is not the most useful, but it is actually pretty hard to get a better one (the `group` parser just doesn't match without a closing brace, so it kicks it back as wholesale invalid, rather than being specific about invalid because the closing brace is missing)

Current feeling is that erroring is better than not, but if someone has an idea of how to get the parser to give a more informative message, I'd be open to hearing it.

**Review by ksunden — APPROVED** [review]

**Comment by ksunden** [comment]
Actually, this diff may get us a better error message:

```diff
diff --git a/lib/matplotlib/_mathtext.py b/lib/matplotlib/_mathtext.py
index 25a825b7b0..de4c25bfe5 100644
--- a/lib/matplotlib/_mathtext.py
+++ b/lib/matplotlib/_mathtext.py
@@ -1894,6 +1894,7 @@ class Parser:
         p.function = csnames("name", self._function_names)
 
         p.group = p.start_group + ZeroOrMore(p.token)("group") + p.end_group
+        p.unclosed_group = p.start_group + ZeroOrMore(p.token)("group") + StringEnd()
 
         p.frac  = cmd(r"\frac", p.required_group("num") + p.required_group("den"))
         p.dfrac = cmd(r"\dfrac", p.required_group("num") + p.required_group("den"))
@@ -1943,6 +1944,7 @@ class Parser:
             p.simple
             | p.auto_delim
             | p.unknown_symbol  # Must be last
+            | p.unclosed_group  # Must be last
         )
 
         p.operatorname = cmd(r"\operatorname", "{" + ZeroOrMore(p.simple)("name") + "}")
@@ -2144,6 +2146,9 @@ class Parser:
     def unknown_symbol(self, s, loc, toks):
         raise ParseFatalException(s, loc, f"Unknown symbol: {toks['name']}")
 
+    def unclosed_group(self, s, loc, toks):
+        raise ParseFatalException(s, len(s), "Expected '}'")
+
     _accent_map = {
         r'hat':            r'\circumflexaccent',
         r'breve':          r'\combiningbreve',
diff --git a/lib/matplotlib/tests/test_mathtext.py b/lib/matplotlib/tests/test_mathtext.py
index d80312495d..5a9e8a8b92 100644
--- a/lib/matplotlib/tests/test_mathtext.py
+++ b/lib/matplotlib/tests/test_mathtext.py
@@ -320,6 +320,7 @@ def test_fontinfo():
         (r'$a^2^2$', r'Double superscript'),
         (r'$a_2_2$', r'Double subscript'),
         (r'$a^2_a^2$', r'Double superscript'),
+        (r'$a = {b$', r"Expected '}'"),
     ],
     ids=[
         'hspace without value',
@@ -347,7 +348,8 @@ def test_fontinfo():
         'unknown symbol',
         'double superscript',
         'double subscript',
-        'super on sub without braces'
+        'super on sub without braces',
+        'unclosed group',
     ]
 )
 def test_mathtext_exceptions(math, msg):
```

@devRD thoughts?

**Comment by devRD** [comment]
I think the diff makes sense, we do get a better error message. Added the changes...

File: lib/matplotlib/_mathtext.py [hunk 1]
```diff
@@ -1865,6 +1865,8 @@ def csnames(group, names):
                 "|".join(map(re.escape, tex2uni)))
         )("sym").leaveWhitespace()
         p.unknown_symbol = Regex(r"\\[A-Za-z]+")("name")
+        def unclosed_group(self, s, loc, toks):
+            raise ParseFatalException(s, len(s), "Expected '}'")
```
Comment (oscargus) [inline lib/matplotlib/_mathtext.py:1869]: The location of this method is not that logical. I'd suggest moving it down to other group-methods. Otherwise, it looks good!

Comment (oscargus) [reply lib/matplotlib/_mathtext.py:1869]: Let me rephrase that: I think this should be removed as there is another method below (at a different level, so I wonder if this is used at all?).

Comment (devRD) [reply lib/matplotlib/_mathtext.py:1869]: I think the method is being utilized because we get the same error message as above. 

Re: _there is another method below (at a different level_ - I might be missing something, could you point out which one?

**Comment by devRD** [comment]
Although not completely sure if this is correct, I think the same behavior is also achieved by redefining the `group` grammar to: 
`p.group = p.start_group + ZeroOrMore(p.token)("group") - p.end_group` 
(The `-` sign here checks for trailing end_groups(`"}"`) and raises an Exception if none is found)

Instead, the error message here would be `ParseSyntaxException: Expected end_group, found end of text`

File: lib/matplotlib/_mathtext.py [hunk 2]
```diff
@@ -2144,6 +2149,9 @@ def symbol(self, s, loc, toks):
     def unknown_symbol(self, s, loc, toks):
         raise ParseFatalException(s, loc, f"Unknown symbol: {toks['name']}")
 
+    def unclosed_group(self, s, loc, toks):
+        raise ParseFatalException(s, len(s), "Expected '}'")
```
Comment (oscargus) [inline lib/matplotlib/_mathtext.py:2153]: Here is the other, identical, method. Should only require one and I think this is the one.

Comment (devRD) [reply lib/matplotlib/_mathtext.py]: Yup, I missed that. Thanks!

File: lib/matplotlib/tests/test_mathtext.py [hunk 3]
```diff
@@ -347,7 +348,8 @@ def test_fontinfo():
         'unknown symbol',
         'double superscript',
         'double subscript',
-        'super on sub without braces'
+        'super on sub without braces',
+        'unclosed group'
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_mathtext.py:352]: So next change doesn't need to modify two lines like this time:
```suggestion
        'unclosed group',
```

File: lib/matplotlib/_mathtext.py [hunk 4]
```diff
@@ -2030,7 +2032,8 @@ def main(self, s, loc, toks):
         return [Hlist(toks)]
 
     def math_string(self, s, loc, toks):
-        return self._math_expression.parseString(toks[0][1:-1])
+        return self._math_expression.parseString(toks[0][1:-1],
+                                                 parseAll=True)
```
Comment (QuLogic) [inline lib/matplotlib/_mathtext.py:2036]: Should fit
```suggestion
        return self._math_expression.parseString(toks[0][1:-1], parseAll=True)
```

**Review by oscargus — APPROVED** [review]
Looks good! Not sure if it should go into 3.8 or 3.9 (I'd vote, always, for 3.8...), so I let one of the reviewers closer to the release process tag and merge.
